### DISCUSSION TITLE
Custom migration, schema snapshot apply and bootstrap results in sort of deadlock

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#9165] [Feature Requests]
Hi all!

I have an interesting predicament:
- I have a custom schema to apply
- I have a custom migration to apply (adds index to a table from custom schema)
- I have an empty database when I start (obviously)

As per [this discussion](https://github.com/directus/directus/discussions/8968) I always run the schema apply first in my container, then the bootstrap CLI command (with the custom migration).

Now in the case of a fresh install, the custom schema cannot be applied (because there are no directus tables yet). Then it goes on to bootstrapping, however, this errors out while running migrations (the table to add the index to doesn't exist yet since the schema wasn't applied). This error results in the fact that the first admin user is never created (because error in the the migrations aren't catched).

Changing the order doesn't fix this problem, because even when I do the bootstrap first, the custom migration still throws an exception resulting in no admin user being created.

Any thoughts on this?

**Comment by rijkvanzanten** [comment]
Very interesting point! Technically, everything here is working as expected, so I made it a discussion for a new way of handling this rather than a bug report 👍🏻 

So the problem here is a chicken/egg problem: You need the migrations to run to apply the schema, and you need the schema applied to run your custom migrations. I suppose the only real way to do that would be to add a flag to skip over custom migrations in bootstrap, and then run those manually after the fact with `database migrate:latest`. 

In the meantime, a workaround could be to add a check to your migration that makes sure the tables exist you're operating on, and then hack around it by undoing and redoing the migration after the schema migration is done, not ideal of course, but should do the trick:

```
// Runs all migrations, including custom ones. Your custom one is a no-op
directus bootstrap

directus apply schema

// Undo the last ran migration (which should be your custom one)
directus database migrate:down

// Now that the schema exists, your custom migration is no longer a no-op
directus database migrate:latest
```

**Reply by cah4a** [reply]
Have the same issue.
What if I have a bunch of custom migrations? A fresh installation is a pain in the ass.

I think the best solution is to add a schema file as an optional parameter to the `bootstrap` cli command.
I could make a PR for that.

**Reply by rijkvanzanten** [reply]
How does that solve the original problem?

**Reply by cah4a** [reply]
> Now in the case of a fresh install, the custom schema cannot be applied (because there are no directus tables yet)

setup directus tables and applying schema before migration should do the trick.

Somewhere around here: https://github.com/directus/directus/blob/main/api/src/cli/commands/bootstrap/index.ts#L22

**Reply by cah4a** [reply]
Also, it makes a lot of sense to me. I want to bootstrap a particular project with all the schema and migrations included.

**Comment by cah4a** [comment]
The thing still hurts.

**Comment by ElGrecode** [comment]
Can someone on the Directus team confirm if this is the expected approach? If so, I could try to submit a PR for it?

The suggested workaround feels really hacky and doesn't account for a dynamic count of migrations.

**Reply by rijkvanzanten** [reply]
> Can someone on the Directus team confirm if this is the expected approach?

Which approach? There's a couple ideas floating around.

> If so, I could try to submit a PR for it?

I'm not quite sure what you'd change 🙂 For new PRs for improvements/features, we do ask you to follow [the contributing guidelines](https://github.com/directus/directus/blob/main/contributing.md#implementing-features). In this case, I'd love it if you could open a new feature request detailing exactly what changes you have in mind, and how you think it should work before starting any development work.

**Reply by ElGrecode** [reply]
> Which approach? There's a couple ideas floating around.


@rijkvanzanten 
 

My suggested approach would be to actually change the sequencing, since this is how most people are using snapshots based on the docs. 

Step 1: Apply base Directus migrations
Step 2: Apply schema snapshot
Step 3: Apply custom migrations

**Reply by rijkvanzanten** [reply]
Mind opening a new feature request detailing that approach? Thanks!

**Reply by ignasigarcia** [reply]
Hey @ElGrecode,

Thanks for sharing this approach about Directus vs Custom migrations.

I am currently facing the same issue and wanted to know if a Proposal has been made because I could not find it.

Cheers

**Reply by ElGrecode** [reply]
@ignasigarcia Didn't make the OS contribution since Directus has started charging licensing fees for their OS project.

I solved it using a sidecar process.

First, this process runs directus bootstrap with an empty migrations folder. No change occurs if DB has migration records.

Then migrations are copied from source. Schema apply. Migrate:latest.

Approach is a little hacky, but a simple solution

**Reply by ignasigarcia** [reply]
Oh that's great to know @ElGrecode!

Thanks for sharing the idea, much appreciated.

Cheers!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
